stan/math/rev/core/vari.hpp

```diff
@@ -300,12 +314,115 @@ class vari_value<T, require_eigen_dense_base_t<T>> : public vari_base {
    */
   inline void init_dependent() { adj_.setOnes(); }
 
+  /**
+   * Implimentation for assignable objects that sets the adjoint values to zero.
+   */
+  template <typename Checker = T,
+            require_t<bool_constant<!std::is_const<Checker>::value>>* = nullptr>
+  inline void set_zero_adjoint_impl() {
+    adj_.setZero();
+  }
+  /**
+   * Implimentation for views which performs a no-op
+   */
+  template <typename Checker = T, require_t<std::is_const<Checker>>* = nullptr>
+  inline void set_zero_adjoint_impl() {}
+
   /**
    * Set the adjoint value of this variable to 0.  This is used to
    * reset adjoints before propagating derivatives again (for
    * example in a Jacobian calculation).
    */
-  inline void set_zero_adjoint() final { adj_.setZero(); }
+  inline void set_zero_adjoint() final { set_zero_adjoint_impl(); }
```

Comment:
```suggestion
  inline void set_zero_adjoint() final { 
    if(!is_constant<T>::value){
      adj_.setZero();
    }
  }
```
We don't need those impls.

Reply:
This actually still fails. If it was `if constexpr (!is_const<T>::value)` this would compile. We could also just make the adjoint part `mutable` so constness is ignored